src/web100srv.c

```diff
@@ -1393,71 +1220,43 @@ void child_process(int parent_pipe, SSL_CTX *ssl_context, int ctlsockfd) {
-                            testopt.connection_flags, JSON_SINGLE_VALUE);
-      break;
-    } else if (parent_message == SRV_QUEUE_SERVER_FAULT || parent_message < 0) {
-      // Send the SRV_FAULT message, exit
-      send_json_message_any(&ctl, SRV_QUEUE, SRV_QUEUE_SERVER_FAULT_STR,
-                            strlen(SRV_QUEUE_SERVER_FAULT_STR),
-                            testopt.connection_flags, JSON_SINGLE_VALUE);
-      exit(0);
-    } else if (parent_message == SRV_QUEUE_SERVER_BUSY) {
-      // Send the SRV_BUSY message, exit
-      send_json_message_any(&ctl, SRV_QUEUE, SRV_QUEUE_SERVER_BUSY_STR,
-                            strlen(SRV_QUEUE_SERVER_BUSY_STR),
-                            testopt.connection_flags, JSON_SINGLE_VALUE);
-      exit(0);
-    } else if (parent_message == SRV_QUEUE_HEARTBEAT) {
-      // Perform the heartbeat check (if you can).  The heartbeat check failing
-      // will cause the client to commit suicide.
-      if (t_opts & TEST_STATUS) {
-        if (send_json_message_any(&ctl, SRV_QUEUE, SRV_QUEUE_HEARTBEAT_STR,
-                                  strlen(SRV_QUEUE_HEARTBEAT_STR),
-                                  testopt.connection_flags,
-                                  JSON_SINGLE_VALUE) != 0) {
-          exit(0);
-        }
-        msg_len = sizeof(buff);
-        if (recv_any_msg(&ctl, &msg_type, buff, &msg_len,
-                         testopt.connection_flags) != 0) {
-          exit(0);
-        }
-      }
-    } else if (parent_message == SRV_QUEUE_SERVER_BUSY_60s) {
-      // Send the 60s message to the client.  The server does not currently
-      // send this message, although web100clt does have code to respond to it.
-      // This message is likely obsolete.
-      if (send_json_message_any(&ctl, SRV_QUEUE, SRV_QUEUE_SERVER_BUSY_60s_STR,
-                                strlen(SRV_QUEUE_SERVER_BUSY_60s_STR),
-                                testopt.connection_flags,
-                                JSON_SINGLE_VALUE) != 0) {
+  while ((retcode = read_from_parent(parent_pipe, &ctl, &testopt, t_opts)) !=
+         SRV_QUEUE_TEST_STARTS_NOW) {
+    if (retcode < 0) {
+      if (retcode == -EINTR) {
+        //  Try to continue an interrupted read.  All other errors are fatal.
+        continue;
+      } else {
         exit(0);
       }
-      alarm(120);
-    } else {
-      // If the number sent down the pipe is positive and not one of the magic
-      // numbers, then it is an estimate of how many minutes the NDT client
-      // will have to wait before its test starts.  Pass that info along.
-      memset(buff, 0, sizeof(buff));
-      sprintf(buff, "%d", parent_message);
-      if (send_json_message_any(&ctl, SRV_QUEUE, buff, strlen(buff),
-                                testopt.connection_flags,
-                                JSON_SINGLE_VALUE) != 0) {
+    }
+    switch (retcode) {
+      case SRV_QUEUE_TEST_STARTS_NOW:
```

Comment:
Hopefully.

It seems like a very good idea to, as much as possible, have a case for every value that an enum can take.  This is an int, but it's used partially as an int and partially as an enum, so I thought it important to call out every "magic number" specified by the protocol.